### ISSUE TITLE
string coords are converted to object dtype when using MultiIndex / stacking

### ISSUE THREAD

**Issue by ctw** [pydata/xarray#2280]
#### Code Sample

```python
import xarray as xr
import pandas as pd

indx = pd.MultiIndex.from_arrays([['1', '2', '1', '2'],
                                  ['100', '100', '200', '200']],
                                 names=['a', 'b'])
# MultiIndex(levels=[['1', '2'], ['100', '200']],
#            labels=[[0, 1, 0, 1], [0, 0, 1, 1]],
#            names=['a', 'b'])

tst = xr.DataArray(data=np.random.random((4)), coords=[('ab', indx)])
# <xarray.DataArray (ab: 4)>
# array([0.754308, 0.268363, 0.799407, 0.513185])
# Coordinates:
#   * ab       (ab) MultiIndex
#   - a        (ab) object '1' '2' '1' '2'
#   - b        (ab) object '100' '100' '200' '200'

tst2 = xr.DataArray(data=np.random.random((2,2)),
                    coords=[('a', ['1', '2']), ('b', ['100', '200'])])
# <xarray.DataArray (a: 2, b: 2)>
# array([[0.854303, 0.179755],
#        [0.463453, 0.478913]])
# Coordinates:
#   * a        (a) <U1 '1' '2'
#   * b        (b) <U3 '100' '200'

tst2.stack(ab=('a', 'b'))
# <xarray.DataArray (ab: 4)>
# array([0.854303, 0.179755, 0.463453, 0.478913])
# Coordinates:
#   * ab       (ab) MultiIndex
#   - a        (ab) object '1' '1' '2' '2'
#   - b        (ab) object '100' '200' '100' '200'
```
#### Problem description

The dtype for `a` and `b` in `tst` and `tst2.stack()` is `object` rather than some string dtype.

#### Expected Output

A string dtype for `a` and `b`.

#### Output of ``xr.show_versions()``

<details>
# Paste the output here xr.show_versions() here
INSTALLED VERSIONS
------------------
commit: None
python: 3.6.5.final.0
python-bits: 64
OS: Linux
OS-release: 4.15.0-24-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

xarray: 0.10.7
pandas: 0.23.1
numpy: 1.14.5
scipy: 1.1.0
netCDF4: 1.4.0
h5netcdf: 0.6.1
h5py: 2.8.0
Nio: None
zarr: None
bottleneck: 1.2.1
cyordereddict: None
dask: 0.18.1
distributed: 1.22.0
matplotlib: 2.2.2
cartopy: None
seaborn: 0.8.1
setuptools: 39.2.0
pip: 9.0.3
conda: 4.5.5
pytest: 3.6.2
IPython: 6.4.0
sphinx: 1.7.5
</details>

**Comment by shoyer** [comment]
Hi @ctw, thanks for raising this issue.

I think we'll be able to fix this eventually, but not until the large refactor envisioned by https://github.com/pydata/xarray/issues/1603.

With the current version of xarray's data model this is basically unavoidable, because pandas.MultiIndex objects hold strings in dtype=object arrays.

**Comment by max-sixty** [comment]
Is this a dupe of https://github.com/pydata/xarray/issues/2280 ?

**Comment by shoyer** [comment]
@max-sixty this *is* #2280 :)

**Comment by max-sixty** [comment]
😂 
https://github.com/pydata/xarray/issues/907 !

**Comment by benbovy** [comment]
This is fixed in v2022.6.0. Xarray's `PandasMultiIndex` wrapper keeps track of the level coordinate dtypes.
